pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/clock"
+)
+
+type backoffRecord struct {
+	key                      string
+	failuresAfterLastSuccess int32
+	lastFailureTime          *time.Time
+}
+
+type backoffStore struct {
+	store cache.Store
+}
+
+func (s *backoffStore) updateBackoffRecord(record backoffRecord) {
+	if b, ok, _ := s.store.GetByKey(record.key); !ok {
+		s.store.Add(&record)
```

Comment:
The `Add` function may error but we silence the error. I think we should propagate it up, or at least log. Same for other functions on store which may error.

WDYT @alculquicondor ?

Reply:
Made changes. PTAL.